src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+# 
+# Next, one employs a smooth threshold projection on the intermediate variable $p_f$ to obtain a "binarized" density parameter $p_t$ that tends towards values of $0$ or $1$ almost everywhere [6]: 
+# ```math
+# p_t = \frac{\tanh(\beta\eta)+\tanh\left[\beta(p_f-\eta)\right]}{\tanh(\beta\eta)+\tanh\left[\beta(1-\eta)\right]}. 
+# ```
+# Note that as $\beta\to\infty$, this threshold procedure goes to a step function, which would make the optimization problem non-differentiable. In consequence, the standard approach is to gradually increase $\beta$ to slowly binarize the design as the optimization progresses [6]. We will show how this is done below.  
+# 
+# Finally, we replace $p$ with the filtered and thresholded new parameter $p_t$ in the material permittivity expression above:
+# 
+# ```math
+# \varepsilon(p_t) = \left[n_{air}+p_t(n_{metal}-n_{air})\right]^2,
+# ```
+# 
+# ## Weak form
+# 
+# Now we get the weak form for the above PDEs. After integral by part and removing the zero boundary integral term, we get:
+# 
+# ```math
+# a(u,v,p) = \int_\Omega \left[\nabla(\Lambda v)\cdot\frac{1}{\varepsilon(p)}\Lambda\nabla u-k^2uv\right]\mathrm{d}\Omega,
+# ```
+# 
+# ```math
+# b(v) = \int_\Omega vf\mathrm{d}\Omega. 
+# ```
+# Notice that the $\nabla(\Lambda v)$ is also a element-wise "product" of two vectors $\nabla$ and $\Lambda v$.
+# 
+
+# ## Setup
+# 
+# We import the packages that will be used, define the geometry and physics parameters. 
+# 
+
+using Gridap, Gridap.Geometry, Gridap.Fields, GridapGmsh
+λ = 532      # Wavelength (nm)
+L = 600      # Width of the numerical cell (excluding PML) (nm)
+h1 = 600     # Height of the air region below the source (nm)
+h2 = 200     # Height of the air region above the source (nm)
+dpml = 300   # Thickness of the PML (nm)
+
+n_metal = 0.054 + 3.429im # Silver refractive index at λ = 532 nm
+n_air = 1    # Air refractive index
+μ = 1        # Magnetic permeability
+k = 2*π/λ    # Wavenumber (nm^-1)
+
+# ## Discrete Model
+# 
+# We import the model from the `RecCirGeometry.msh` mesh file using the `GmshDiscreteModel` function defined in `GridapGmsh`. The mesh file is created with GMSH in Julia (see the file ../assets/TopOptEMFocus/MeshGenerator.jl). Note that this mesh file already contains periodic boundary information for the left and right side, and that is enough for gridap to realize a periodic boundary condition should be implemented. Also, the center smallest distance circle region is labeled with `Center` and the "donut" design region is labeled with `Design`. 
+# 
+
+model = GmshDiscreteModel("../models/RecCirGeometry.msh")
+
+# ## FE spaces for the magnetic field
+# 
+# We use the first-order lagrangian as the finite element function space basis. The dirihlet edges are labeld with `DirichletEdges` in the mesh file. Since our problem involves complex numbers (because of PML), we need to assign the `vector_type` to be `Vector{ComplexF64}`.
```

Comment:
```suggestion
# We use the first-order Lagrange finite-element basis functions. The Dirichlet edges are labeled as `DirichletEdges` in the mesh file. Since our problem involves complex numbers (because of the PML), we need to specify the `vector_type` as `Vector{ComplexF64}`.
```